swim/handlers_test.go

```diff
@@ -0,0 +1,222 @@
+
+func (s *HandlerTestSuite) TearDownTest() {
+	s.ctxCancel()
+
+	// Destroy any nodes added to the peer list during testing
+	for _, node := range s.peers {
+		if node != nil {
+			node.Destroy()
+		}
+	}
+	if s.cluster != nil {
+		s.cluster.Destroy()
+	}
+}
+
+func (s *HandlerTestSuite) TestGossipStartHandler() {
+	s.testNode.node.gossip.Stop()
+	s.Require().True(s.testNode.node.gossip.Stopped())
+
+	_, err := s.testNode.node.gossipHandlerStart(s.ctx, &arg{})
+	s.NoError(err, "calling handler should not result in error")
+
+	s.False(s.testNode.node.gossip.Stopped())
+}
+
+func (s *HandlerTestSuite) TestGossipStopHandler() {
+	s.testNode.node.gossip.Start()
+	s.Require().False(s.testNode.node.gossip.Stopped())
+
+	_, err := s.testNode.node.gossipHandlerStop(s.ctx, &arg{})
+	s.NoError(err, "calling handler should not result in error")
+
+	s.True(s.testNode.node.gossip.Stopped())
+}
+
+func (s *HandlerTestSuite) TestToggleGossipHandler() {
+	s.Require().True(s.testNode.node.gossip.Stopped())
+
+	var err error
+
+	// Toggle gossip on
+	_, err = s.testNode.node.gossipHandler(s.ctx, &arg{})
+	s.NoError(err, "calling handler should not result in error")
+	s.False(s.testNode.node.gossip.Stopped())
+
+	// Toggle gossip off
+	_, err = s.testNode.node.gossipHandler(s.ctx, &arg{})
+	s.NoError(err, "calling handler should not result in error")
+	s.True(s.testNode.node.gossip.Stopped())
+}
+
+func (s *HandlerTestSuite) TestAdminLeaveJoinHandlers() {
+	// Join test node to cluster
+	s.cluster.Add(s.testNode.node)
+	s.Require().Equal(5, s.testNode.node.CountReachableMembers(), "expect a cluster of 5")
+
+	var err error
+	var status *Status
+
+	// Test leave handler works correctly
+	status, err = s.testNode.node.adminLeaveHandler(s.ctx, &arg{})
+	s.NoError(err, "calling handler should not result in error")
+	s.Equal(&Status{Status: "ok"}, status)
+	s.Equal(4, s.testNode.node.CountReachableMembers())
+
+	// Test join handler brings it back to 5
+	status, err = s.testNode.node.adminJoinHandler(s.ctx, &arg{})
+	s.NoError(err, "calling handler should not result in error")
+	s.Equal(&Status{Status: "rejoined"}, status)
+	s.Equal(5, s.testNode.node.CountReachableMembers())
+}
+
+// TestRegisterHandlers tests that registerHandler always succeeds.
+func (s *HandlerTestSuite) TestRegisterHandlers() {
+	s.NoError(s.testNode.node.registerHandlers())
+}
+
+func (s *HandlerTestSuite) TestNotImplementedHandler() {
+	_, err := notImplementedHandler(s.ctx, &arg{})
+	s.EqualError(err, "handler not implemented")
+}
+
+// TestErrorHandler tests that the errorHandler logs the correct error message.
+func (s *HandlerTestSuite) TestErrorHandler() {
+	logger := &mocks.Logger{}
+
+	errTest := errors.New("test error")
+
+	logger.On("WithField", "error", errTest).Return(logger)
+	logger.On("Info", []interface{}{"error occurred"})
+
+	originalLogger := s.testNode.node.log
+	s.testNode.node.log = logger
+
+	s.testNode.node.errorHandler(s.ctx, errTest)
+	logger.AssertExpectations(s.T())
```

Comment:
Based on our conversation yesterday, I focused on adding test coverage for significant branches. I stopped trying to test all the branches that were merely logging.

However, this test has already been written so I figure it might as well stay. Do you agree?